lib/Mojo/IOLoop/Stream/Client.pm

```diff
@@ -0,0 +1,127 @@
+package Mojo::IOLoop::Stream::Client;
+use Mojo::Base 'Mojo::IOLoop::Stream';
+
+use Mojo::Util 'term_escape';
+
+use constant DEBUG => $ENV{MOJO_CLIENT_DEBUG} || 0;
+
+sub new {
+  my $self = shift->SUPER::new(@_);
+  $self->on(read => sub { shift->_read_content(shift) });
+  $self->on(close => sub { $_[0]->{closing}++ || $_[0]->_finish(1) });
+  return $self;
+}
+
+sub process {}
+
+sub _error {
+  my ($self, $err) = @_;
+  $self->{tx}->res->error({message => $err}) if $self->{tx};
```

Comment:
The private property `$self->{tx}` is only assigned by subclasses in their overloaded `process` method. That breaks encapsulation.